packages/expo-screen-orientation/ios/EXScreenOrientation/EXScreenOrientationModule.m

```diff
@@ -0,0 +1,222 @@
+{
+    UIInterfaceOrientationMask orientationMask = [self getSupportedInterfaceOrientations];
+    NSDictionary *maskToOrienationMap = @{
+        @(UIInterfaceOrientationMaskPortrait): @(UIInterfaceOrientationPortrait),
+        @(UIInterfaceOrientationMaskPortraitUpsideDown): @(UIInterfaceOrientationPortraitUpsideDown),
+        @(UIInterfaceOrientationMaskLandscapeLeft): @(UIInterfaceOrientationLandscapeLeft),
+        @(UIInterfaceOrientationMaskLandscapeRight): @(UIInterfaceOrientationLandscapeRight)
+    };
+    // If the particular orientation is supported, we add it to the array of allowedOrientations
+    NSMutableArray *allowedOrientations = [[NSMutableArray alloc] init];
+    for (NSNumber *wrappedSingleOrientation in [maskToOrienationMap allKeys]) {
+      UIInterfaceOrientationMask supportedOrientationMask = orientationMask & [wrappedSingleOrientation intValue];
+      if (supportedOrientationMask) {
+        UIInterfaceOrientation supportedOrientation = [maskToOrienationMap[@(supportedOrientationMask)] intValue];
+        [allowedOrientations addObject:[EXScreenOrientationUtilities exportOrientation:supportedOrientation]];
+      }
+    }
+    resolve(allowedOrientations);
+}
+
+UM_EXPORT_METHOD_AS(supportsOrientationLockAsync,
+                    supportsOrientationLockAsync:(NSNumber *)orientationLock
+                    resolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  UIInterfaceOrientationMask orientationMask = [EXScreenOrientationUtilities importOrientationLock:orientationLock];
+  if (!orientationMask) {
+    resolve(@NO);
+  } else if ([EXScreenOrientationUtilities doesSupportOrientationMask:orientationMask]) {
+    resolve(@YES);
+  } else {
+    resolve(@NO);
+  }
+}
+
+UM_EXPORT_METHOD_AS(getOrientationAsync,
+                    getOrientationAsyncResolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  resolve([EXScreenOrientationUtilities exportOrientation:_currentScreenOrientation]);
+}
+
+// Will be called when this module's first listener is added.
+- (void)startObserving
+{
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(handleDeviceOrientationChange:)
+                                               name:UIDeviceOrientationDidChangeNotification
+                                             object:[UIDevice currentDevice]];
+  _hasListeners = YES;
+}
+
+// Will be called when this module's last listener is removed, or on dealloc.
+- (void)stopObserving
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+  _hasListeners = NO;
+}
+
+- (void)handleDeviceOrientationChange:(NSNotification *)notification
+{
+  UIDevice *device = notification.object;
+  UIInterfaceOrientation currentDeviceOrientation = [EXScreenOrientationUtilities UIDeviceOrientationToUIInterfaceOrientation:[device orientation]];
```

Comment:
Let's not rely on `device.orientation` when calculating the interface orientation — as per [the documentation](https://developer.apple.com/documentation/uikit/uidevice/1620053-orientation?language=objc):

> This value represents the physical orientation of the device and **may be different** from the current orientation of your application’s user interface.

We should probably keep on using `traitCollectionDidChange:` on `UIViewController` which will inform us of trait changes on `UIViewController`, as per [the documentation](https://developer.apple.com/documentation/uikit/uitraitenvironment/1623516-traitcollectiondidchange?language=objc):

> The system calls this method when the iOS interface environment changes. Implement this method in view controllers and views, according to your app’s needs, to respond to such changes. For example, you might adjust the layout of the subviews of a view controller when an iPhone is rotated from portrait to landscape orientation.

One thing I would want to check would be to see if rotating 10.5" iPad triggers the method too. Basing on the image:
<img width="758" alt="Zrzut ekranu 2019-12-17 o 11 24 40" src="https://user-images.githubusercontent.com/1151041/70987274-d4d82980-20bf-11ea-856d-49b515cc06ad.png">
traits in both orientations are the same, so maybe the method wouldn't be called?

Reply:
On 10.5'' iPad, `traitCollectionDidChange` isn't called when device orientation was changed. 
Moreover, based on trait collection we only can distinguish between `landspace` and `portraint`. Calculating orientation from device orientation and the current mask will give us more information. 

The user has only two way to change screen orientation: rotate the device or change it using our API. When he uses our API, `ScreenOrienationModule` simulates physical rotation. When he rotates the device, `UIDeviceOrientationDidChangeNotification` will be fired. So we catch this event in `ScreenOrientationModule`.

Reply:
So basically, depending on the device orientation fixes a bug that would result in screen orientation callback not being called when rotating the 10.5" iPad, right? Superb! 🎉